### PR TITLE
pallet_authority_discovery: introduce current_authorities and next_authorities methods 

### PR DESCRIPTION
**Pull request by ordian**
In polkadot we need to map ValidatorId to AuthorityIds. The way we did it is assume the same ordering of keys stored with OneSessionHandlers of the corresponding keys. However in the current implementation the authority discovery sorts its keys.

This PR introduces pallet_authority_discovery::current_authorities method which preserves the ordering.

polkadot companion: https://github.com/paritytech/polkadot/pull/2262

### CODE REVIEW

File: client/authority-discovery/src/worker.rs [hunk 1]
```diff
@@ -328,18 +328,30 @@ where
...
 			.map_err(Error::CallingRuntime)?
 			.into_iter()
-			.filter(|id| !local_keys.contains(id.as_ref()))
-			.collect();
+			.filter(|id| !local_keys.contains(id.as_ref()));
+		let next_authorities = self
+			.client
+			.runtime_api()
+			.next_authorities(&id)
+			.unwrap_or_default()
```
Comment (ordian) [inline client/authority-discovery/src/worker.rs:342]: the runtime error here is ignored, I don't know whether we will need that when introduced on a live chain or whether we will need a migration

File: frame/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -74,15 +82,15 @@ impl<T: Config> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
...
+	fn on_new_session<'a, I: 'a>(_changed: bool, validators: I, queued_validators: I)
 	where
 		I: Iterator<Item = (&'a T::AccountId, Self::Key)>,
 	{
 		// Remember who the authorities are for the new and next session.
-		if changed {
-			let keys = validators.chain(queued_validators).map(|x| x.1).collect::<BTreeSet<_>>();
-			Keys::put(keys.into_iter().collect::<Vec<_>>());
-		}
+		let keys = validators.map(|x| x.1).collect::<Vec<_>>();
```
Comment (bkchr) [inline frame/authority-discovery/src/lib.rs:90]: Why is the check on changed removed?

File: primitives/authority-discovery/src/lib.rs [hunk 3]
```diff
@@ -47,7 +47,9 @@ sp_api::decl_runtime_apis! {
 	/// This api is used by the `client/authority-discovery` module to retrieve identifiers
 	/// of the current and next authority set.
 	pub trait AuthorityDiscoveryApi {
-		/// Retrieve authority identifiers of the current and next authority set.
+		/// Retrieve authority identifiers of the current authority set.
 		fn authorities() -> Vec<AuthorityId>;
+		/// Retrieve authority identifiers of the next authority set.
+		fn next_authorities() -> Vec<AuthorityId>;
```
Comment (bkchr) [inline primitives/authority-discovery/src/lib.rs:53]: Please bump the version and write a migration.

Comment (bkchr) [reply primitives/authority-discovery/src/lib.rs:53]: Or we keep just one method as this runtime api is authority specific and that only cares about all session keys.

No reason to have two methods.

Comment (bkchr) [reply primitives/authority-discovery/src/lib.rs:53]: We only need the change on the runtime side.

File: frame/authority-discovery/src/lib.rs (hunk 2 above)
Comment (ordian) [reply frame/authority-discovery/src/lib.rs:90]: Does it apply to the current session only? I.e. do we have to store `NextKeys` irregardless?

File: primitives/authority-discovery/src/lib.rs (hunk 3 above)
Comment (rphmeier) [reply primitives/authority-discovery/src/lib.rs:53]: @ordian I agree with Basti - the `AuthorityDiscovery::authorities()` function can retain the same behavior of combining authorities from multiple validator sets. This is fine as long as there is a way for internal runtime code to get the actual current authorities with correct sorting.

**Review by rphmeier — APPROVED** [review]
Looks like the storage of `Keys` will be wrong until the next session after the runtime upgrade. This seems minor for our use-cases in Polkadot and the behavior of the `AuthorityDiscovery` runtime API should be the same as before. I don't believe we need a storage migration as such.

File: frame/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -80,9 +96,11 @@ impl<T: Config> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
 	{
 		// Remember who the authorities are for the new and next session.
 		if changed {
-			let keys = validators.chain(queued_validators).map(|x| x.1).collect::<BTreeSet<_>>();
-			Keys::put(keys.into_iter().collect::<Vec<_>>());
+			let keys = validators.map(|x| x.1);
+			Keys::put(keys.collect::<Vec<_>>());
 		}
+		let keys = queued_validators.map(|x| x.1);
```
Comment (bkchr) [inline frame/authority-discovery/src/lib.rs:102]: Please put it into the if. You can read the what changed is doing in the docs.

File: frame/authority-discovery/src/lib.rs [hunk 5]
```diff
@@ -47,15 +49,29 @@ decl_module! {
...
+		let next = NextKeys::get();
+
+		keys.extend(next);
+		keys.sort();
+		keys.dedup();
+
+		keys
+	}
+
+	/// Retrieve authority identifiers of the current authority set in the original order.
```
Comment (bkchr) [inline frame/authority-discovery/src/lib.rs:65]: We should probably also provide one for next.

**Review by bkchr — APPROVED** [review]
One last thing needs to be done.

File: frame/authority-discovery/src/lib.rs [hunk 6]
```diff
@@ -18,12 +18,12 @@
 //! # Authority discovery module.
 //!
 //! This module is used by the `client/authority-discovery` to retrieve the
```
Comment (mxinden) [inline frame/authority-discovery/src/lib.rs:20]: If I am not mistaken `current_authorities` and `next_authorities` would be called by the parachain logic. I would suggest either amending the comment or removing it all together.

**Review by mxinden — APPROVED** [review]
Approving from a `client/authority-discovery` perspective. For what my review on a runtime change is worth, this looks good to me. Thanks @ordian.

**Comment by ordian** [comment]
@bkchr do I need to bump `pallet_authority_discovery` version?

**Comment by bkchr** [comment]
> @bkchr do I need to bump `pallet_authority_discovery` version?

No is fine.

**Comment by ordian** [comment]
bot merge

**Pull request merged**